### PR TITLE
Use a nonce

### PR DESCRIPTION
**Pull request by jglick**
As per [spec](https://openid.net/specs/openid-connect-core-1_0.html#NonceNotes). Not sure how important this is in practice. Not tested yet against any realistic provider.

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/oic/OicSession.java [hunk 1]
```diff
@@ -54,7 +55,13 @@
     /**
      * An opaque value used by the client to maintain state between the request and callback.
      */
-    private final String state = Base64.encode(UUID.randomUUID().toString().getBytes(StandardCharsets.UTF_8)).substring(0,20);
+    @VisibleForTesting
+    String state = Base64.encode(UUID.randomUUID().toString().getBytes(StandardCharsets.UTF_8)).substring(0,20);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/oic/OicSession.java:59]: Unclear to my this was not just
```suggestion
    String state = UUID.randomUUID().toString();
```
but not touching that here.

File: src/test/java/org/jenkinsci/plugins/oic/PluginTest.java [hunk 2]
```diff
@@ -326,8 +324,7 @@ private String createIdToken(PrivateKey privateKey, Map<String, Object> keyValue
             .setIssuer("issuer")
             .setSubject(TEST_USER_USERNAME)
             .setAudience(Collections.singletonList("clientId"))
-            .setAudience(System.currentTimeMillis() / 60 + 5)
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/oic/PluginTest.java:329]: Was this supposed to be the `iat`?

File: src/test/java/org/jenkinsci/plugins/oic/PluginTest.java [hunk 3]
```diff
@@ -326,8 +324,7 @@ private String createIdToken(PrivateKey privateKey, Map<String, Object> keyValue
...
-            .setIssuedAtTimeSeconds(System.currentTimeMillis() / 60);
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/oic/PluginTest.java:330]: Did not smell right. (Why 60?) Anyway unnecessary in this test.

File: src/test/java/org/jenkinsci/plugins/oic/TestRealm.java [hunk 4]
```diff
@@ -51,13 +50,9 @@ public TestRealm(WireMockRule wireMockRule, String userInfoServerUrl, String ema
...
-        try {
-            Field field = OicSession.class.getDeclaredField("state");
-            field.setAccessible(true);
-            field.set(OicSession.getCurrent(), "state");
-        } catch (Exception e) {
-            throw new RuntimeException("can't fudge state",e);
-        }
+        OicSession session = OicSession.getCurrent();
+        session.state = "state";
+        session.nonce = "nonce";
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/oic/TestRealm.java:55]: No need to use reflection in the same package.

**Comment by jglick** [comment]
I can resolve merge conflicts if there is some indication someone is going to look at this. CC @Skiepp

**Comment by Skiepp** [comment]
@jglick I just updated the plugin for being version compatible and restore properly the group permission
I'm not planning on performing updates and I do not have anymore permission

Feel free to adopt this plugin

**Comment by michael-doubez** [comment]
@jglick I am updating the plugin. I should be ready to merge features sometime next week or the week after.

Nonce parameter is high on my list 😊

**Comment by michael-doubez** [comment]
LGTM.

Regarding the impacts, I cannot guess.
Nonce is pretty standard in OIDC providers (and in some case required) but a non-compliant one could fail to propagate it back.

I'll add a special note in the release.

**Review by michael-doubez — APPROVED** [review]

**Comment by jglick** [comment]
I presume you will do some sanity checks against major providers before release, since as I mentioned I have not tried this patch live.

**Comment by michael-doubez** [comment]
@jglick yes, will do it. I'd like to release it with PR regarding the parsing of groups which is usually a headache.

**Comment by michael-doubez** [comment]
I will merge the PR and then add an option to disable nonce usage (and whatever modification is needed after tests).

**Pull request merged**